### PR TITLE
Make gateway POST handle multipart file & pin flag

### LINKED ISSUE #1819: gateway: file upload functionality?
Would it make sense to enable adding a file to node via HTTP POST/PUT request to gateway? This would enable IPFS to be used in lots of web applications that just need GET/PUT functionality. This option should not probably be default and could be enabled via a command line switch (similarly how unrestricted-api is enabled).

### PR DESCRIPTION
**Pull request by rht**
An updated #1845 
- [x] tests
- [x] fixes #1819

### CODE REVIEW

**Comment by rht** [comment]
Test fixed. https://github.com/rht/go-ipfs/commit/666b755926ffdd1efe28f98bafb6e6cb89aa3351#diff-cbe32203deec9b995950e3110e5db668R69 is an example of adding an array of files (could be in the form of `file[]=...`, and can be nested).

Several apis to look at, in the order of decreasing features:
- http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html (max 5G or 5TB with multipart upload; has encryption)
- https://developers.google.com/drive/v2/reference/files/insert (max size 5TB; has revision; has ocr)
- https://www.dropbox.com/developers-v1/core/docs#files_put (max size 150 MB or 48 hrs with /chunked_upload; has revisions)

None of them support upload of multiple files in a single POST request.
The aws [cli](http://aws.amazon.com/cli/), however, has the sync command for folders.
What would be the reason for supporting file arrays?

imo, further dedup with `ipfs add` should eventually be done. Currently, the remaining differences with `ipfs add` are the params (Hidden, Trickle, Wrap, Chunker), and that the latter has a chan to output the hash of each files.

File: commands/files/multipartfile.go [hunk 1]
```diff
@@ -55,6 +55,22 @@ func NewFileFromPart(part *multipart.Part) (File, error) {
 	return f, nil
 }
 
+func NewFileFromRequest(r *http.Request) (File, error) {
+	contentType := r.Header.Get(contentTypeHeader)
+	mediaType, _, _ := mime.ParseMediaType(contentType)
```
Comment (rht) [inline commands/files/multipartfile.go]: err ignored due to https://github.com/ipfs/go-ipfs/commit/ce49541f130e736639ff270ad0132245f2938ee8. If there is a more semantic way to express that this is run only when files are involved.

Comment (rht) [inline commands/files/multipartfile.go]: redirect: https://github.com/ipfs/go-ipfs/issues/2226

**Comment by jbenet** [comment]
odd, travis didn't run... @rht can you rebase + push again so travis runs? thanks

File: commands/files/multipartfile.go (hunk 1 above)
Comment (jbenet) [inline commands/files/multipartfile.go]: i dont understand this statement.

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -287,21 +289,57 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
+		nd, err := fileAdder.Finalize()
+		if err != nil {
+			return "", err
+		}
+		k, err := nd.Key()
+		if err != nil {
+			return "", err
+		}
+
+		dopin := r.FormValue("pin")
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: shouldnt use the form value, this check should check the command option.

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -287,21 +289,57 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
+	k, err := addAllAndPin(f)
 	if err != nil {
 		internalWebError(w, err)
 		return
 	}
 
 	i.addUserHeaders(w) // ok, _now_ write user's headers.
-	w.Header().Set("IPFS-Hash", k.String())
-	http.Redirect(w, r, ipfsPathPrefix+k.String(), http.StatusCreated)
+	w.Header().Set("IPFS-Hash", k)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: why remove `.String()` ?

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: ah nvm there is no command, it's gateway

File: test/sharness/t0111-gateway-writeable.sh [hunk 4]
```diff
@@ -39,6 +39,46 @@ test_expect_success "We can HTTP GET file just created" '
...
+  HASH=$(echo $LOCATION | cut -d":" -f2- |tr -d " \n\r")
+'
+
+test_expect_success "We can ipfs get the files just created" '
+  #ipfs get is used here instead since curl does not download multiple files
+  ipfs get -o actual "$HASH" &&
+  test_cmp infile1 actual/infile1 &&
+  test_cmp infile2 actual/infile2
+'
+
```
Comment (jbenet) [inline test/sharness/t0111-gateway-writeable.sh]: nice test :+1:

File: core/corehttp/gateway_handler.go (hunk 3 above)
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: i'm assuming its because it gets called implicitly somewhere

File: core/corehttp/gateway_handler.go [hunk 5]
```diff
@@ -287,21 +289,57 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
+			return "", err
+		}
+
+		dopin := r.FormValue("pin")
+		if dopin == "" || dopin == "true" {
+			if err := fileAdder.PinRoot(); err != nil {
+				return "", err
+			}
+		}
+		return k.String(), nil
```
Comment (rht) [inline core/corehttp/gateway_handler.go]: Here is the implicit `.String()` call.

**Comment by rht** [comment]
Rebased.

**Comment by whyrusleeping** [comment]
so, just to make sure i understand the goal here. You're wanting to use the unixfs fileAdder for writeable gateway code?

**Comment by rht** [comment]
Yes, this dedup has been intended since #1543.

File: core/corehttp/gateway_handler.go [hunk 6]
```diff
@@ -301,21 +303,57 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 }
 
 func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
-	nd, err := i.newDagFromReader(r.Body)
+	fileAdder, err := coreunix.NewAdder(i.node.Context(), i.node, nil)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: we should use the closenotify code like in the get or head requests to make sure we don't have lingering uncanceled contexts from here.

**Comment by whyrusleeping** [comment]
just one context thing and then LGTM.

**Comment by rht** [comment]
Added the context closenotifer to all gateway requests.

**Pull request closed without merging**